sklearn/metrics/ranking.py

```diff
@@ -172,16 +184,57 @@ def average_precision_score(y_true, y_score, average="macro",
     >>> y_true = np.array([0, 0, 1, 1])
     >>> y_scores = np.array([0.1, 0.4, 0.35, 0.8])
     >>> average_precision_score(y_true, y_scores)  # doctest: +ELLIPSIS
-    0.79...
+    0.83...
+    >>> import numpy as np
+    >>> from sklearn.metrics import average_precision_score
+    >>> yt = np.array([0, 0, 1, 1])
+    >>> ys = np.array([0.1, 0.4, 0.35, 0.8])
+    >>> ap = average_precision_score(yt, ys, interpolation='eleven_point')
+    >>> ap # doctest: +ELLIPSIS
+    0.84...
 
     """
-    def _binary_average_precision(y_true, y_score, sample_weight=None):
+    def _binary_uninterpolated_average_precision(
+            y_true, y_score, sample_weight=None):
         precision, recall, thresholds = precision_recall_curve(
             y_true, y_score, sample_weight=sample_weight)
-        return auc(recall, precision)
+        return -sum(np.diff(recall) * np.array(precision)[:-1])
+
+    def _binary_eleven_point_average_precision(
+            y_true, y_score, sample_weight=None):
+        precision, recall, thresholds = precision_recall_curve(
+            y_true, y_score)
+
+        # We need the recall values in ascending order
+        precision = list(reversed(precision))
+        recall = list(reversed(recall))
+
+        eleven_precisions = list()
+        # Don't use the first (precision, recall) pair with precision = 1.
+        start = 1
+        for target_recall in np.arange(0, 1.1, 0.1):
```

Comment:
Can't you do this with something along the lines of `precision[numpy.searchsorted(recall, np.arange(0, 1.1, 0.1))]`?